### ISSUE TITLE
Error in conversion from void* to GumboNode*

### ISSUE THREAD

**Issue by dedSea** [google/gumbo-parser#318]
I'm receiving the following error while trying to find the title from within a page being passed in as a char* by curl.

    Data_Handles.h:169:43: error:
    invalid conversion from ‘void*’ to ‘GumboNode* {aka GumboInternalNode*}’ [-fpermissive]
    GumboNode* child = root_children->data[i];

I receive the same error in another location, but seeing how the code is similar, I figure that it'd be better to post once.

The following is (with a small change) the same code as within the `get_titles.c` file in the examples directory of the repository.

    155:	char* toParse = get_src(url);
    158:	GumboOutput* output = gumbo_parse(toParse);
    161:	GumboNode* root = (output->root);
    163:	assert(root->type == GUMBO_NODE_ELEMENT);
    164:	assert(root->v.element.children.length >= 2);

    166:	const GumboVector* root_children = &root->v.element.children;
    167:	GumboNode* head = NULL;
    168:	for (int i = 0; i < root_children->length; ++i) {
    169:		GumboNode* child = root_children->data[i];
    170:		if (child->type == GUMBO_NODE_ELEMENT &&
    171:			child->v.element.tag == GUMBO_TAG_HEAD) {
    172:			head = child;
    173:			break;
    172:		}
    173:	}
    174:	assert(head != NULL);

I apologize for not being able to narrow the error down any further.   I have tested the output of the curl and it does pass in valid source.  I have used the `clean_text.cc` example file as a test for that.  Therefore, I assume it's an issue with the way I'm using the structure of the GumboOutput and GumboNode.

Thank you in advance for any help!

**Comment by nostrademons** [comment]
Are you compiling as C (gcc, .c extension) or C++ (g++, .cc extension)?  Implicit casts from void* to any other pointer type are legal in C, but they aren't legal in C++.  Use 'GumboNode* child = static_cast<GumboNode*> root_children->data[i]' instead.  There are examples in the unit tests, which are written in C++.

**Comment by dedSea** [comment]
I'm compiling `.cpp` files with g++.

**Comment by nostrademons** [comment]
Try using static_cast\<GumboNode*> then.

**Comment by dedSea** [comment]
That removed the errors.  Once I correct few errors within my program and verify that did the trick, I'll let you know.
Thank you!

**Comment by dedSea** [comment]
Using those allowed my program to successfully build.  But now, when executing the built program, I'm receiving:

    Test: src/tokenizer.c:561: emit_current_tag: Assertion `output->original_text.data[0] == '<'' failed.
    Aborted

The title-getting feature is the only one used in this program (so far).

**Comment by nostrademons** [comment]
What's the input you're running it on, does it parse with the get_title example, and what's the code you're using to invoke Gumbo?  Also which version of the library are you using?  An assertion failure shouldn't show up in normal operation.

**Comment by dedSea** [comment]
I cloned the repo with git and installed it as listed on the README.

I have not run the unit tests because I'm receiving errors that I cannot resolve with CMake.  Google-ing the error hasn't brought any clarity on the matter.

I use the following code to invoke Gumbo:

    string parse(string toFind){
		if(toFind == "title"){
			return find_title(const_cast<char*>(landingSite.c_str()));
		}	
	}

Since `std::string.c_str()` returns `const char*`, I used `const_cast<char*>` to pass the url into the the function because it only takes in char*.  The function then gets the source from the url via curl.

I will verify the return of the source from curl and get back to you.

**Comment by dedSea** [comment]
The curl line does return valid source from the website being tested.  In this case, it's a wikipedia article:

http://en.wikipedia.org/wiki/Imprint_%28TV_series%29

The <title> tag contains: Imprint (TV series) - Wikipedia, the free encyclopedia

Could this problem be caused by the fact that %28 and %29 both change (in the title and in the URL) into parentheses?

**Comment by nostrademons** [comment]
It's possible but unlikely that it would affect Gumbo.

I would save the HTML source that you're getting back from curl to a file, and then run one of the example programs on it.  That will tell you whether it's a problem in your usage of Gumbo or whether you've uncovered a bug in the library.  FWIW, when I saved that page from my browser and ran get_title on it it ran fine, so it sounds like something either may be misformatted coming out of curl or you may be running into object lifetime issues.

**Comment by dedSea** [comment]
I'll give it a go.

But when testing the same way (with curl usage) with the clean_text.cc program, I didn't have an issue.  I'll give the original source a try with reading from a file, but ultimately, I'd like to be able to do it from the web.

Another update shortly.  Thanks for bearing with me.

**Comment by dedSea** [comment]
Awesome, it returns the correct value!  I'm going to be trying again with the input from curl again tonight, but thank you so much for your help in getting my C++ code working!

**Comment by dedSea** [comment]
Just to give you an update, with your help, I solved the problem.  I ended up using the exact code from `get_title.c` with `static_cast<GumboNode*>` preceding code to get data from children nodes.  For example:

`GumboNode* child = root_children->data[i];` became `GumboNode* child = static_cast<GumboNode*>(root_children->data[i]);`

Instead of storing my source to a file and increasing time for I/O, I wanted to put my source directly into a `char*` and pass it into the `gumbo_parse_with_options()` function.  When passing the source directly into the function, I received the error:

    Test: src/tokenizer.c:561: emit_current_tag:     
    Assertion `output->original_text.dta[0] == '<'' failed.`

I figured this means there was an invalid token.  To check, I stripped off all characters whose ASCII values are less than 0 or greater than 128.

After doing so, I obtained the length of the string (and added 1 to account for null characters), and passed the values into the function.

The program compiled and executed without issue (except some pointer-freeing nonsense).

The resulting code is as follows:

	const char* parse(){
		char* toParse = get_src_curl(const_cast<char*>(scrapeURL.c_str()));

		string s(toParse);
		stripUnicode(s);

		char* strippedSource = const_cast<char*>(s.c_str());
		int input_length = strlen(strippedSource) + 1;

		GumboOutput* output = gumbo_parse_with_options(
			&kGumboDefaultOptions, strippedSource, input_length);

		const char* title = find_title(output->root);

		gumbo_destroy_output(&kGumboDefaultOptions, output);
		free(toParse);
		free(strippedSource);

		return title;
	}

Thank you very much for your help and for the development of such a such a speedy parser!
